### PR TITLE
DI-326 Replace event bridge

### PR DESCRIPTION
**Pull request by ed-sparkes**
## Link to JIRA Ticket

https://nhsd-jira.digital.nhs.uk/browse/DI-326

## Description
EventBridge API Destinations does not behave as we were expecting and as such we need to replace it. 

We need a different solution that allows us to throttle how we send change requests to the DoS API Gateway. 

We have opted for an orchestrator pattern as described by Jeremy Daly here  https://www.jeremydaly.com/throttling-third-party-api-calls-with-aws-lambda/

This solution provides resilience against DoS being down or throttled. 

It also ensures messages are delivered in order. 

A detailed write up will follow in the ADR but here is the architecture diagram for reference

image

### Noteworthy Changes

- All the changes in this PR are noteworthy

## Type of change

Delete not appropriate

- Bug fix (non-breaking change which fixes an issue)

## Testing

Please tick the testing that has been completed

- [x] Unit tests
- [ ] Integration tests

## Developer Checklist

- [x] I have performed a self-review of my own code
- [x] I have run the [code formatting checks](../README.md#code-quality)
- [x] I have run the [code quality checks](../README.md#code-quality)
- [x] New code meets [standards](https://nhsd-confluence.digital.nhs.uk/display/DI/DI+Ways+of+Working) agreed by the team
- [x] Unit test code coverage is at or above 80%
- [x] New and existing unit tests pass locally with my changes
- [ ] Tests have added that prove my fix is effective or that my feature works (Integration tests)
- [ ] I have made corresponding changes to the documentation
- [ ] I have cleaned down my environment (if created)

## Code Reviewer Checklist

- [ ] I have run the unit tests and they run correctly
- [ ] I can confirm the changes have been tested or approved by a tester
- [ ] I can confirm no remaining infrastructure is left over from this branch

### CODE REVIEW

File: application/event_sender/event_sender.py [hunk 1]
```diff
@@ -78,23 +81,25 @@ def lambda_handler(event: ChangeRequestQueueItem, context: LambdaContext, metric
...
-                #         "ods_code": {"DataType": "String", "StringValue": self.nhs_entity.odscode},
-                #     },
-                #      )
+                logger.info("Permanent error sending to DLQ, Not retrying")
+                hashed_payload = sha256(str(event["change_request"]).encode()).hexdigest()
+                sqs.send_message(
+                    QueueUrl=environ["CR_DLQ_URL"],
+                    MessageBody=dumps(event["change_request"]),
+                    MessageDeduplicationId=f"{time_ns()}-{hashed_payload}",
+                    MessageGroupId=odscode,
```
Comment (ed-sparkes) [inline application/event_sender/event_sender.py:93]: have used serviceId here see event_processor that puts the original message on the queue

Comment (ed-sparkes) [reply application/event_sender/event_sender.py:93]: i.e. add  MessageDeduplicationId and MessageGroupId to the metadata

File: application/event_sender/tests/test_event_sender.py [hunk 2]
```diff
@@ -135,7 +139,7 @@ def test_lambda_handler_dos_api_fail(
     mock_put_metric.assert_has_calls([call("DosApiLatency", 0, "Milliseconds"), call("DoSApiFail", 1, "Count")])
     assert response["statusCode"] == 500
     assert response["body"] == "something went wrong"
-    mock_client.return_value.delete_message.asset_not_called()
```
Comment (ed-sparkes) [inline application/event_sender/tests/test_event_sender.py]: good spot, this is a bit of an issue, with no type checking its easy to miss this as the test would have passed even if the function had been called

File: application/common/dynamodb.py [hunk 3]
```diff
@@ -20,6 +20,47 @@ def dict_hash(change_event: Dict[str, Any], sequence_number: str) -> str:
...
+        serializer = TypeSerializer()
+        put_item = {k: serializer.serialize(v) for k, v in dynamo_record.items()}
+        response = dynamodb.put_item(TableName=environ["CHANGE_EVENTS_TABLE_NAME"], Item=put_item)
+        logger.info("Put circuit status", extra={"response": response, "item": put_item})
+    except Exception as err:
+        logger.exception(f"Unable to insert a record into dynamodb.Error: {err}")
+        raise
+
+
+def get_circuit_is_open(circuit: str) -> Union[bool, None]:
```
Comment (mabe13) [inline application/common/dynamodb.py:40]: Function could do with a description block like "add_change_request_to_dynamodb", without context not sure what the purpose of getting this open stat is for

File: infrastructure/stacks/cloudwatch-dashboard/main.tf [hunk 4]
```diff
@@ -242,7 +247,7 @@ resource "aws_cloudwatch_dashboard" "cloudwatch_dashboard" {
             "height": 6,
             "properties": {
                 "metrics": [
-                    [ "AWS/SQS", "NumberOfMessagesReceived", "QueueName", "${var.dead_letter_queue_from_event_bus_name}", { "label": "EventBridge Message Count" } ],
+                    [ "AWS/SQS", "NumberOfMessagesReceived", "QueueName", "${var.cr_dead_letter_queue_from_fifo_queue_name}", { "label": "EventBridge Message Count" } ],
```
Comment (rohith1122) [inline infrastructure/stacks/cloudwatch-dashboard/main.tf:250]: rename EventBridge to cr_fifo

File: test/integration/steps/test_parent_steps.py [hunk 5]
```diff
@@ -386,7 +339,7 @@ def event_bridge_dlq_log_check(context):
         f'fields message | sort @timestamp asc | filter correlation_id="{context["correlation_id"]}"'
         ' | filter report_key="EVENTBRIDGE_DLQ_HANDLER_RECEIVED_EVENT"'
```
Comment (rohith1122) [inline test/integration/steps/test_parent_steps.py:340]: Rename EVENTBRIDGE to CR_FIFO

File: test/integration/steps/test_parent_steps.py [hunk 6]
```diff
@@ -386,7 +339,7 @@ def event_bridge_dlq_log_check(context):
...
     )
-    logs = get_logs(query, "eb_dlq", context["start_time"])
+    logs = get_logs(query, "cr_dlq", context["start_time"])
     assert (
         "Eventbridge Dead Letter Queue Handler received event" in logs
```
Comment (rohith1122) [inline test/integration/steps/test_parent_steps.py:344]: Rename Eventbridge to cr_fifo

**Review by rohith1122 — DISMISSED** [review]
LGTM

File: application/event_sender/event_sender.py [hunk 7]
```diff
@@ -1,68 +1,104 @@
...
 logger = Logger()
 
 
 @tracer.capture_lambda_handler()
 @unhandled_exception_logging
-@validate_event_is_signed
 @logger.inject_lambda_context
 @metric_scope
-def lambda_handler(event: Dict[str, Any], context: LambdaContext, metrics) -> Dict:
+def lambda_handler(event: ChangeRequestQueueItem, context: LambdaContext, metrics) -> Dict:
```
Comment (mabe13) [inline application/event_sender/event_sender.py:23]: This function could do with some refactoring but that can be done in a new ticket later

**Review by rohith1122 — APPROVED** [review]
LGTM

**Review by mabe13 — APPROVED** [review]
Good Stuff 👍 
Could do with some refactoring in places (event sender) which can be done in a separate ticket later
Merge when ready ✅

File: application/event_sender/event_sender.py (hunk 7 above)
Comment (ed-sparkes) [reply application/event_sender/event_sender.py:23]: ok will review in separate ticket

**Pull request merged**